vendors/nordic/boards/nrf52840-dk/ports/common_io/iot_spi.c

```diff
@@ -0,0 +1,598 @@
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+/*
+ * Event handler passed to NRF SPI controller. Used for sync and async modes.
+ * Note, pvContext is set to IotSPIHandle_t of corresponding SPI instance at open.
+ */
+void prvSpiEventHandler( nrf_drv_spi_evt_t const * pxEvent,
+                         void * pvContext )
+{
+    IotSPIHandle_t xHandle = ( IotSPIHandle_t ) pvContext;
+    const SpiContext_t * pxSpi = xHandle->pxSpiContext;
+
+    switch( pxEvent->type )
+    {
+        case NRF_DRV_SPI_EVENT_DONE:
+            /* Adjust atomic flags. These are used by sync modes for polling and reproduces internal nRF_SDK blocking behaviour */
+            CRITICAL_REGION_ENTER();
+            bTransferDone = true;
+            CRITICAL_REGION_EXIT();
+
+            /* In case of async transfers, event handlers should have been set, and should now be invoked */
+            if( xHandle->xSpiCallback != NULL )
+            {
+                xHandle->xSpiCallback( eSPISuccess, xHandle->pvUserContext );
+            }
+
+            break;
+
+        default:
+            break;
+    }
+}
+
+/*
+ * Transfer logic, with less parameter checking.
+ * Maintains internal atomics and applies appropriate CommonIO error codes
+ * Called by other CommonIO that check input parameters before calling.
+ */
+static int32_t prvSpiTransfer( IotSPIHandle_t const pxSPIPeripheral,
+                               uint8_t const * pucTxBuffer,
+                               size_t xTxSize,
+                               uint8_t const * pucRxBuffer,
+                               size_t xRxSize,
+                               bool bSync )
+{
+    int32_t lError = IOT_SPI_SUCCESS;
+
+    if( bTransferDone )
+    {
+        IotSPICallback_t pxUserCallback;
+        SpiContext_t * pxSpiContext = pxSPIPeripheral->pxSpiContext;
+
+        /* Less code by just making callback NULL while in sync transfer. i.e. no callback
+         * This is fine because it's not possible for any transaction to fire if any transaction, sync or async, is underway.
+         */
+        if( bSync )
+        {
+            pxUserCallback = pxSPIPeripheral->xSpiCallback;
+            pxSPIPeripheral->xSpiCallback = NULL;
+        }
+
+        /* There was no transfer in progress, so we can safely attempt a new transfer */
+        CRITICAL_REGION_ENTER();
+        bTransferDone = false;
+        CRITICAL_REGION_EXIT();
+
+        ret_code_t xTransferStatus = nrf_drv_spi_transfer( &pxSpiContext->instance, pucTxBuffer, xTxSize, pucRxBuffer, xRxSize );
+
+        if( NRF_SUCCESS != xTransferStatus )
+        {
+            CRITICAL_REGION_ENTER();
+            bTransferDone = true;
+            CRITICAL_REGION_EXIT();
+
+            lError = IOT_SPI_TRANSFER_ERROR;
+        }
+        else
+        {
+            if( bSync )
+            {
+                /* For sync transfers, we block until they complete */
+                while( !bTransferDone )
```

Comment:
If you want it non-blocking, you can use a semaphore as an alternative. You take the semaphore here and you give the semaphore in the event handler.

If you want it blocking, it seems you can pass a NULL event handler to nrf_drv_spi_transfer, in which case it is blocking mode.

Reply:
I did consider using semaphore, but was basing of ST CommonIO where only vendor HAL is used. Also implemented as such to mimic Nordic [driver blocking/non-blocking implementation](https://github.com/aws/amazon-freertos/blob/master/vendors/nordic/nRF5_SDK_15.2.0/modules/nrfx/drivers/src/nrfx_spim.c#L523) which busy-polls an event register.

`nrf_drv_spi_transfer` doesn't take an event handler as an argument. At initialization, `nrf_drv_spi_init` takes an event handler and if it's NULL, then subsequent calls to `nrf_drv_spi_transfer` for that SPI instance will be blocking and it it's non-NULL then subsequent transfers will be async with the input callback. Given that, it's more work to constantly init/unit with and without callbacks just to have the HAL drivers internally implement blocking/non-blocking. Hence the permanent installation of `prvSpiEventCallback` which handles sync/async behaviour without having to init/uninit the SPI instance so frequently, and does so in a similar manner that [the HAL driver does it](https://github.com/aws/amazon-freertos/blob/master/vendors/nordic/nRF5_SDK_15.2.0/modules/nrfx/drivers/src/nrfx_spim.c#L523).